### PR TITLE
Introduce `EmbeddedQuicChannel`

### PR DESCRIPTION
**Pull request by NiteshKant**
__Motivation__

Today tests mock a `QuicChannel`. Over time the mocking is becoming more verbose as we require more features in http/3, eg: while adding QPACK dynamic tables, `createStream()` are required to be implemented. As `EmbeddedChannel` exists and extending it to support `QuicStreamChannel` is not much work, it is less tedious to implement and use `EmbeddedQuicChannel` instead of the mocks.

__Modification__

- Implemented `EmbeddedQuicChannel`
- Modified all usages of mocks for the `QuicChannel` to now use `EmbeddedQuicChannel`

__Result__

Less tedious changes to tests (mocks) when implementing new http/3 features.

### CODE REVIEW

File: src/test/java/io/netty/incubator/codec/http3/EmbeddedQuicChannel.java [hunk 1]
```diff
@@ -0,0 +1,275 @@
...
+    public void peerAllowedStreams(QuicStreamType type, long peerAllowedStreams) {
+        this.peerAllowedStreams.put(type, peerAllowedStreams);
+    }
+
+    @Override
+    public Future<QuicStreamChannel> createStream(QuicStreamType type, ChannelHandler handler,
+                                                  Promise<QuicStreamChannel> promise) {
+        promise.setSuccess(new EmbeddedQuicStreamChannel(this, true, type,
+                streamIdUpdater.getAndAdd(this, 2), handler));
+        return promise;
```
Comment (normanmaurer) [inline src/test/java/io/netty/incubator/codec/http3/EmbeddedQuicChannel.java:101]: nit: you could merge the above lines as `promise.setSuccess` returns the promise as well

File: src/test/java/io/netty/incubator/codec/http3/EmbeddedQuicChannel.java [hunk 2]
```diff
@@ -0,0 +1,275 @@
...
+        } else {
+            reason.release();
+        }
+        return close(promise);
+    }
+
+    @Override
+    public Future<QuicConnectionStats> collectStats(Promise<QuicConnectionStats> promise) {
+        promise.setFailure(new UnsupportedOperationException("Collect stats not supported for embedded channel."));
+        return promise;
```
Comment (normanmaurer) [inline src/test/java/io/netty/incubator/codec/http3/EmbeddedQuicChannel.java:118]: nit: you could merge the above lines as `promise.setSuccess` returns the promise as well

File: src/test/java/io/netty/incubator/codec/http3/Http3ControlStreamOutboundHandlerTest.java [hunk 3]
```diff
@@ -57,20 +52,31 @@
...
 
+    @Override
+    public void testValidTypeInbound() {
+        super.testValidTypeInbound();
+    }
+
+    @Override
+    public void testValidTypeOutput() {
+        super.testValidTypeOutput();
+    }
```
Comment (normanmaurer) [inline src/test/java/io/netty/incubator/codec/http3/Http3ControlStreamOutboundHandlerTest.java:63]: Why is this needed ?

**Review by normanmaurer — APPROVED** [review]
This looks great! I had 3 small nits but feel free to ignore

File: src/test/java/io/netty/incubator/codec/http3/Http3ControlStreamOutboundHandlerTest.java (hunk 3 above)
Comment (NiteshKant) [reply src/test/java/io/netty/incubator/codec/http3/Http3ControlStreamOutboundHandlerTest.java:63]: Yikes debugging leftover, will remove.

File: src/test/java/io/netty/incubator/codec/http3/AbtractHttp3ConnectionHandlerTest.java [hunk 4]
```diff
@@ -40,11 +40,16 @@ public void testOpenLocalControlStream() throws Exception {
...
+        final EmbeddedQuicStreamChannel localControlStream = quicChannel.localControlStream();
+        assertNotNull(localControlStream);
+
         assertNotNull(Http3.getLocalControlStream(quicChannel));
 
         handler.channelInactive(ctx);
         handler.channelUnregistered(ctx);
         handler.handlerRemoved(ctx);
+
+        localControlStream.finishAndReleaseAll();
```
Comment (normanmaurer) [inline src/test/java/io/netty/incubator/codec/http3/AbtractHttp3ConnectionHandlerTest.java:52]: you may also want to assert the return value.. True for all calls :)

Comment (NiteshKant) [reply src/test/java/io/netty/incubator/codec/http3/AbtractHttp3ConnectionHandlerTest.java:52]: Fixed in #141
